slru/slru.go

```diff
@@ -0,0 +1,140 @@
+package slru
+
+import (
+	"container/list"
+)
+
+type segment int
+
+const (
+	probation segment = iota
+	protected
+)
+
+// Cache is a segmented LRU cache. It is not safe for concurrent access.
+type Cache struct {
+	items        map[string]*list.Element
+	probation    *list.List
+	protected    *list.List
+	maxProbation int
+	maxProtected int
+}
+
+type entry struct {
+	key     []byte
+	value   []byte
+	segment segment
+}
+
+// New creates a new SLRU cache.
+//
+// Segment capacities must be positive.
+func New(maxProbation int, maxProtected int) *Cache {
+	if maxProbation < 1 || maxProtected < 1 {
+		panic("slru: segment capacities must be positive")
+	}
+
+	return &Cache{
+		items:        make(map[string]*list.Element),
+		probation:    list.New(),
+		maxProbation: maxProbation,
+		protected:    list.New(),
+		maxProtected: maxProtected,
+	}
+}
+
+// Get looks up a key's value from the cache.
+func (c *Cache) Get(key []byte) (value []byte, ok bool) {
+	keyStr := string(key)
+	e, hit := c.items[keyStr]
+	if !hit {
+		return
+	}
+	ent := e.Value.(*entry)
+	value, ok = ent.value, true
+
+	if ent.segment == protected {
+		c.protected.MoveToFront(e)
+		return
+	}
+
+	// Just promote the entry there's room in the next segment.
+	if c.protected.Len() < c.maxProtected {
+		c.probation.Remove(e)
+		ent.segment = protected
+		c.items[keyStr] = c.protected.PushFront(ent)
+		return
+	}
+
+	// Swap the entry with the oldest protected one in-place to minimize allocations.
+
+	prot := c.protected.Back()
+	victim := prot.Value.(*entry)
+	victim.segment = probation
+	ent.segment = protected
+	prot.Value, e.Value = e.Value, prot.Value
+
+	c.protected.MoveToFront(prot)
+	c.probation.MoveToFront(e)
+	c.items[keyStr] = prot
+	c.items[string(victim.key)] = e
+	return
+}
+
+// Add adds a value to the cache.
+func (c *Cache) Add(key []byte, value []byte) {
+	keyStr := string(key)
+	if _, ok := c.items[keyStr]; ok {
+		panic("slru: replacement of existing keys is not supported")
```

Comment:
This edge case could be handled, but I was unsure what correct behavior would be. I chose to be explicit about not handling it.  If this were to pass unchecked, the cache would leak capacity with each duplicate add.